src/platform/Zephyr/KeyValueStoreManagerImpl.cpp

```diff
@@ -0,0 +1,117 @@
+#include <lib/support/ReturnMacros.h>
+#include <settings/settings.h>
+
+#include <support/logging/CHIPLogging.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_DECLARE(chip);
+
+namespace chip {
+namespace DeviceLayer {
+namespace PersistedStorage {
+
+KeyValueStoreManagerImpl KeyValueStoreManagerImpl::sInstance;
+
+void KeyValueStoreManagerImpl::Init()
+{
+    int result = settings_subsys_init();
+
+    if (result != 0)
+        ChipLogError(DeviceLayer, "settings_subsys_init() failed with error: %d", result);
+}
+
+struct ReadEntry
+{
+    void * destination;            // NOTE: can be nullptr in which case `readSize` should still be returned
+    size_t destinationBufferSize;  // size of destination buffer
+    size_t readSize;               // [out] size of read entry value
+    CHIP_ERROR result;             // [out] read result
+};
+
+int LoadEntryCallback(const char * name, size_t entrySize, settings_read_cb readCb, void * cbArg, void * param)
+{
+    // If requested a key X, process just node X and ignore all its descendants: X/*
+    if (settings_name_next(name, nullptr) > 0)
+        return 0;
+
+    ReadEntry & entry = *reinterpret_cast<ReadEntry *>(param);
+
+    if (!entry.destination || entrySize > entry.destinationBufferSize)
+    {
+        entry.result   = CHIP_ERROR_BUFFER_TOO_SMALL;
+        entry.readSize = entrySize;
+        return 0;
+    }
+
+    // Found requested key
+    const ssize_t bytesRead = readCb(cbArg, entry.destination, entry.destinationBufferSize);
+    entry.result            = bytesRead > 0 ? CHIP_NO_ERROR : CHIP_ERROR_PERSISTED_STORAGE_FAILED;
+    entry.readSize          = bytesRead > 0 ? bytesRead : 0;
+    return 0;
+}
+
+CHIP_ERROR KeyValueStoreManagerImpl::_Get(const char * key, void * value, size_t value_size, size_t * read_bytes_size,
+                                          size_t offset_bytes) const
+{
+    VerifyOrReturnError(key, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(value, CHIP_ERROR_INVALID_ARGUMENT);
+    // Offset and partial reads are not supported, for now just return NOT_IMPLEMENTED. Support can be added in the
+    // future if this is needed.
+    VerifyOrReturnError(offset_bytes == 0, CHIP_ERROR_NOT_IMPLEMENTED);
+
+    ReadEntry entry{ value, value_size, 0, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND };
+    settings_load_subtree_direct(key, LoadEntryCallback, &entry);
+
+    // Assign readSize only in case read_bytes_size is not nullptr, as it is optional argument
+    if (read_bytes_size)
+        *read_bytes_size = entry.readSize;
+
+    return entry.result;
+}
+
+CHIP_ERROR KeyValueStoreManagerImpl::_Put(const char * key, const void * value, size_t value_size)
+{
+    VerifyOrReturnError(key, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(value, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(value_size > 0, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(settings_save_one(key, value, value_size) == 0, CHIP_ERROR_PERSISTED_STORAGE_FAILED);
```

Comment:
Unless `key` have some very CHIP specific content, we should consider introducing the root path e.g. "ch/". This is common practice in Thread, BLE and many other components in Zephyr e.g.:
 - https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/openthread/platform/settings.c#L16
 - https://github.com/zephyrproject-rtos/zephyr/blob/1342d12500867d3de09025037e65e756d5486eba/subsys/bluetooth/host/settings.c#L211
 - https://github.com/nrfconnect/sdk-nrfxlib/blob/b35fc6728b5565b0d7046511df561f8151d7f106/nrf_security/src/psa/psa_eits_zephyr_settings.c#L18

Apart of better separation, i believe the `DeleteAll` API may be introduced soon, and such prefix will help to wipe all CHIP specific data.

Reply:
But do you propose adding it in the platform-specific code? Since now all platforms will support string-based keys maybe it would be possible to apply that suggestion to the common CHIP code?